README.md
# pg_gnufind
GNU find output as foreign tables. Requires python 3.3+.

## Install
On an Ubuntu system, run the following:

```bash
  $ sudo apt-get update
```

If not already installed, install PostgreSQL and required modules. Note: this FDW requires Python 3, so be sure
you install the python3 version of multicorn.

```bash
  $ sudo apt-get install postgresql-9.6 postgresql-9.6-python3-multicorn
```
Download pg_geekspeak and run the following inside the project directory

```bash
  $ ./tests.py
```

Verify all tests pass, then install.

```bash
  $ sudo dpkg -i pg-gnufind_1.0.0-1_all.deb
```

Within PostgreSQL, install the multicorn extension, create the foreign data server, and make your tables.

```sql
  -- Add the extension. This is a hard dependency. It will not work without it.
  CREATE EXTENSION multicorn;

  -- Create the foreign server. This is not a server in the traditional TCP/IP sense. This is a mapping
  -- between PostgreSQL's foreign data wrapper interface and any foreign tables you may create.
  CREATE SERVER gnufind  -- This can be any name you like, but it's best to be descriptive.
    FOREIGN DATA WRAPPER multicorn
    OPTIONS (wrapper 'ttfkam.FindWrapper');  -- Searches the default python path

  -- Now create your table. All of these columns are optional.
  -- Note: "group" and "user" are quoted because they are reserved SQL words.
  CREATE FOREIGN TABLE gs.media_fdw (
    accessed timestamptz NOT NULL,  -- Last accessed (reverts to last modified if noatime is set)
    changed timestamptz NOT NULL,   -- Last changed (reverts to last modified if noctime is set)
    depth smallint NOT NULL,        -- Levels deep inside the search root
    dirname varchar,                -- Just the directory portion, omitting file
    eperms varchar NOT NULL,        -- Permissions in expanded form
    filename varchar,               -- Just the file portion, omitting directories
    filesystem varchar,             -- Filesystem type, e.g., ext4, zfs
    gid int4 NOT NULL,              -- Filesystem entry group id (see: /etc/group)
    "group" varchar NOT NULL,       -- Filesystem entry group name (see: /etc/group)
    hardlinks smallint NOT NULL,    -- Number of hardlinks that refer to this bag o' bytes
    inum int8 NOT NULL,             -- inode number from the filesystem
    modified timestamptz NOT NULL,  -- Last modified
    path varchar NOT NULL,          -- Path relative to the search root
    perms varchar NOT NULL,         -- Permissions in octal form
    size bigint NOT NULL,           -- Storage space used
    symlink character varying,      -- If it's a symbolic link, where it points to
    type character(1) NOT NULL,     -- Entry type, e.g., 'f' for file, 'd' for directory
    uid int4 NOT NULL,              -- Filesystem entry user id (see: /etc/passwd)
    "user" varchar NOT NULL,        -- Filesystem entry user name (see: /etc/passwd)

    -- Here's where it gets fun. Warning, accessing external program output hurts performance
    mime varchar,
    encoding varchar,

    -- Debugging. If you have a problem with WHERE clauses, I'll need this data to fix it.
    debug_quals text
  )
  SERVER gnufind  -- Make sure this matches your CREATE SERVER statement above
  OPTIONS (
    -- This option is mandatory and the directory needs to exist.
    root_directory '/var/some/dir/to/scan/',

    -- This is how the mime and encoding are gathered as listed above.
    -- You can pass in any program as long as it returns only a single line of text.
    mime '/usr/bin/file -L -b --mime-type',
    encoding '/usr/bin/file -L -b --mime-encoding'
  );
```



debian/changelog
pg-gnufind (1.0.0-1) experimental; urgency=low

  * Initial release.

 -- Miles Elam <miles@geekspeak.org>  Fri, 23 Jun 2017 21:49:16 -0700

tests.py
#!/usr/bin/env python3

from unittest import TestCase, main
from unittest.mock import patch, MagicMock
from io import StringIO

class Qual:
  def __init__(self, name, op, value, is_list=False):
    self.field_name = name
    self.operator = op
    self.value = value
    self.is_list_operator = is_list

class ForeignDataWrapper:
  def __init__(self, options, columns):
    pass

class multicorn:
  ForeignDataWrapper = ForeignDataWrapper

import sys
sys.modules['multicorn'] = multicorn

from gnufind import FindWrapper

ID_BUILTIN = 0
ID_PATTERN = 1
ID_EXECUTABLE = 2

TEST_DIR = '/usr/share/example/'
TEST_ARGS = ['/usr/bin/find', '-O3', TEST_DIR,
             '-regextype', 'posix-egrep', '-ignore_readdir_race']

class FindWrapperTests(TestCase):

  def test_root_with_slash(self):
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': '/dir/no/slash/' }, columns)
    self.assertEqual(fw._root, '/dir/no/slash/')

  def test_root_without_slash(self):
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': '/dir/no/slash' }, columns)
    self.assertEqual(fw._root, '/dir/no/slash/')

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_basic(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\t2017-03-22+22:33:15.3646792370\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR }, columns)
    self.assertIn('path', fw._handlers)
    self.assertEqual(fw._handlers['path'][0], ID_BUILTIN)
    self.assertEqual(fw._handlers['path'][1], '%P')
    self.assertIn('modified', fw._handlers)
    self.assertEqual(fw._handlers['modified'][0], ID_BUILTIN)
    self.assertEqual(fw._handlers['modified'][1], '%T+')
    row = next(fw.execute([], columns))
    expected = TEST_ARGS + [ '-printf', '%P\t%T+\n' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe, universal_newlines=True)
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:15.3646792370', 'path': 'example.txt' })

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_multi(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example1.txt\t2017-03-22+22:33:15.3646792370\n' +
                                         'example2.txt\t2017-03-22+22:33:45.3646792370\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR }, columns)
    rows = fw.execute([], columns)
    row = next(rows)
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:15.3646792370', 'path': 'example1.txt' })
    row = next(rows)
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:45.3646792370', 'path': 'example2.txt' })

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_exec(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\t2017-03-22+22:33:15.3646792370\n' +
                                         'text/plain\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'modified': None, 'mime_type': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR,
                       'mime_type': '/usr/bin/file -b -i' }, columns)
    row = next(fw.execute([], columns))
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:15.3646792370',
                            'path': 'example.txt',
                            'mime_type': 'text/plain' })
    expected = TEST_ARGS + [ '-printf', '%P\t%T+\n',
                             '-exec', '/usr/bin/file', '-b', '-i', '{}', ';' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe, universal_newlines=True)

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_alias(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\t2017-03-22+22:33:15.3646792370\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'ts': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR,
                       'ts': 'modified' }, columns)
    row = next(fw.execute([], columns))
    self.assertEqual(row, { 'ts': '2017-03-22 22:33:15.3646792370',
                            'path': 'example.txt' })
    expected = TEST_ARGS + [ '-printf', '%P\t%T+\n' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe, universal_newlines=True)

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_quals(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'depth': None, 'type': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR }, columns)
    row = next(fw.execute([Qual('depth','=',2), Qual('type','=','f')], { 'path': None }))
    self.assertEqual(row, { 'path': 'example.txt' })
    expected = TEST_ARGS + [ '-mindepth', '2', '-maxdepth', '2', '-type', 'f', '-printf', '%P\n' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe, universal_newlines=True)

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_pattern(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO(TEST_DIR + 'example.txt\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'stem': None, 'extension': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR,
                       'stem': '(?P<stem>[^.]+)\\.(?P<extension>[^.]+)' }, columns)
    row = next(fw.execute([], columns))
    self.assertEqual(row, { 'extension': 'txt', 'stem': 'example' })
    expected = TEST_ARGS + [ '-printf', '%P\n' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe, universal_newlines=True)

  @patch('gnufind.Popen')
  @patch('gnufind.PIPE')
  def test_anonymous_pattern(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO(TEST_DIR + 'example.txt\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'stem': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR,
                       'stem': '([^.]+)\\.[^.]+' }, columns)
    row = next(fw.execute([], columns))
    self.assertEqual(row, { 'stem': 'example' })
    expected = TEST_ARGS + [ '-printf', '%P\n' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe, universal_newlines=True)

if __name__ == '__main__':
  main()


FindWrapper.py
from multicorn import ForeignDataWrapper
import re
from subprocess import Popen, PIPE

# wrapper 'geekspeak.FindWrapper',
# options(
#   prefix='/usr/bin/head -c 10 {}'    # first ten bytes
#   suffix='/usr/bin/tail -c 10 {}'    # last ten bytes
#   mimetype='/usr/bin/file -b -i {}'  # mime type
#   season='s(?P<season>\\d{2})e(?P<episode>\\d{2})(?:  -  (?P<name>.*?))\.(?P<extension>[^.]{3,4})'
# )
class FindWrapper(ForeignDataWrapper):

  def __init__(self, options, columns):
    super(FindWrapper, self).__init__(options, columns)
    self._root = FindWrapper.__extract_root_directory(options)
    self._handlers = {}
    self._patterns = []
    FindWrapper.__init_handlers(self._handlers, self._patterns, options, columns)

  def execute(self, quals, columns):
    handlers = [
      [],  # builtins
      [],  # patterns
      [],  # extensions
    ]
    path_index = -1
    for colname in columns:  # organize our columns into query types, e.g., patterns
      handler = self._handlers[colname]
      handlers[handler[0]].append((colname, handler[1], handler[2], handler[3]))
      if colname == 'path' and handler[0] == 0:  # if we've got a path and it's not aliased
        path_index = len(handlers[0])  # track it for patterns

    builtins = list(map((lambda h: h[1]), handlers[0]))  # build up our list of patterns

    patterns = {}  # patterns need a path
    if len(handlers[1]) > 0:
      for p in handlers[1]:
        pattern_id = p[1]
        patterns[pattern_id] = self._patterns[pattern_id]
      if path_index == -1:  # otherwise, add it to the query patterns
        builtins.append('%P')
        path_index = len(handlers[0])

    # set up program arguments
    args = ['/usr/bin/find', '-O3', '-ignore_readdir_race', self._root]

    for qual in quals:  # process quals to reduce raw find output
      args += self._handlers[qual.name][3](qual) or []

    args += [ '-printf', US.join(builtins) + '\n' ]  # append query patterns to program args

    extensions = list(map((lambda h: h[1]), handlers[2]))  # set up extension queries
    num_exts = len(extensions)
    for extension in extensions:
      args += ['-exec'] + extension.split(' ') + ['{}', ';']

    proc = Popen(args, stdout=PIPE)  # run the program
    for line in proc.stdout:  # …and get the results
      row = {}

      ext_handlers = handlers[2]  # process extensions
      for i in range(0, len(ext_handlers)):
        colname = ext_handlers[i][0]
        # each -exec comes on its own new line
        row[colname] = proc.stdout.readline().rstrip("\n\r")

      builtin_handlers = handlers[0]  # process builtins
      parts = line.rstrip("\n\r").split(US)
      for i in range(0, len(builtin_handlers)):
        colname = builtin_handlers[i][0]
        row[colname] = builtin_handlers[i][3](parts[i])

      matches = {}  # process patterns
      for i in range(0, len(patterns)):
        matches.update(patterns[i].fullmatch(parts[path_index]).groupdict())
      for colname in list(map((lambda h: h[0]), handlers[1])):
        row[colname] = matches[colname]
      yield row
    proc.close()
    return None

  @staticmethod
  def __extract_root_directory(options):
    if 'root_directory' in options:
      root = options['root_directory']
      del options['root_directory']
      if root[-1:] != '/':
        root += '/'
      return root
    else:
      log_to_postgres(logging.ERROR, 'No root_directory specified in options')

  @staticmethod
  def __init_handlers(handlers, patterns, options, columns):
    for option in options:
      if option not in columns:
        log_to_postgres(logging.ERROR, 'Invalid column: ' + option)
        return
      if option in handlers:
        log_to_postgres(logging.ERROR, 'Column option defined more than once: ' + option)
        return
      value = options[option]
      if value[0] in ('/', '~'):  # executable
        handlers[option] = (2, value, noop_qual, default_serializer)
      elif option.find('(') == -1:
        if value in BUILTINS:
          alias = BUILTINS[value]
          serializer = default_serializer if len(alias) < 3 else alias[2]
          handlers[option] = (0, alias[0], alias[1], serializer)
        else:
          log_to_postgres(logging.ERROR, 'Invalid alias: ' + value)
          return
      else:
        colnames = [m.group(1) for m in PATTERN_RE.finditer(value) if m]
        if option in colnames:
          for colname in colnames:
            if colname in handlers:
              log_to_postgres(logging.ERROR, 'Column defined more than once: ' + value)
              return
            handlers[colname] = (1, len(patterns), noop_qual, default_serializer)
          try:
            patterns.append(re.compile(value))
          except re.error:
            log_to_postgres(logging.ERROR, 'Invalid pattern: ' + value)
            return
        else:
          log_to_postgres(logging.ERROR, 'Invalid column: ' + option)
          return
    for colname in columns:
      if colname in handlers:
        continue  # Already handled by aliasing
      elif colname in BUILTINS:
        handler = BUILTINS[colname]
        serializer = default_serializer if len(handler) < 3 else handler[2]
        handlers[colname] = (0, handler[0], handler[1], serializer)
      else:
        log_to_postgres(logging.ERROR, 'Invalid column: ' + colname)
        return


PATTERN_RE = re.compile('\(\?<\s*([^ \)]+)\s*>')

def time_serialize(val):
  return val.replace('+', ' ')

def default_serializer(val):
  return val

def time_qual(name):
  def q(qual):
    if qual.operator == '=':
      return ['-' + name + 'min', '0']
    elif qual.operator in ('<', '<='):
      return ['-not', '-newer' + name + 't', qual.value]
    elif qual.operator in ('>', '>='):
      return ['-newer' + name + 't', qual.value]
  return q

def num_qual(name):
  def q(qual):
    if qual.operator == '=':
      return ['-' + name, qual.value]
  return q

def name_qual(qual):
  if qual.operator == '~~':
    return ['-name', qual.value.replace('%', '*')]
  elif qual.operator == '~~*':
    return ['-iname', qual.value.replace('%', '*')]
  elif qual.operator == '!~~':
    return ['-not', '-name', qual.value.replace('%', '*')]
  elif qual.operator == '!~~*':
    return ['-not', '-iname', qual.value.replace('%', '*')]

def depth_qual(qual):
  if qual.operator == '=':
    return ['-depth', qual.value]
  elif qual.operator == '!=':
    return ['-not', '-depth', qual.value]
  elif qual.operator == '<':
    return ['-maxdepth', qual.value]
  elif qual.operator == '>':
    return ['-mindepth', qual.value]
  elif qual.operator == '<=':
    return ['-maxdepth', qual.value + 1]
  elif qual.operator == '>=':
    return ['-mindepth', qual.value - 1]

def dir_qual(qual):
  return ['-true']

def fs_qual(name):
  if qual.operator == '=':
    return ['-fstype', qual.value]
  elif qual.operator == '!=':
    return ['-not', '-fstype', qual.value]

def hardlink_qual(qual):
  if qual.operator == '=':
    return ['-links', qual.value]
  elif qual.operator == '<':
    return ['-links', '-' + qual.value]
  elif qual.operator == '>':
    return ['-links', '+' + qual.value]
  elif qual.operator == '<=':
    return ['-links', '-' + (qual.value + 1)]
  elif qual.operator == '>=':
    return ['-links', '+' + (qual.value - 1)]

def symlink_qual(qual):
  if qual.operator == '~~':
    return ['-lname', qual.value.replace('%', '*')]
  elif qual.operator == '~~*':
    return ['-ilname', qual.value.replace('%', '*')]
  elif qual.operator == '!~~':
    return ['-not', '-lname', qual.value.replace('%', '*')]
  elif qual.operator == '!~~*':
    return ['-not', '-ilname', qual.value.replace('%', '*')]

def path_qual(qual):
  return ['-true']

def perm_qual(qual):
  return ['-true']

def type_qual(qual):
  if qual.operator == '=':
    return ['-type', qual.value]

def owner_qual(name):
  def q(qual):
    if qual.operator == '=':
      return ['-' + name, qual.value]
  return q

def size_qual(qual):
  if qual.operator == '=':
    return ['-size', qual.value + 'c']
  elif qual.operator == '<':
    return ['-size', '-' + qual.value + 'c']
  elif qual.operator == '>':
    return ['-size', '+' + (qual.value + 1) + 'c']
  elif qual.operator == '<=':
    return ['-size', '-' + (qual.value + 1) + 'c']
  elif qual.operator == '>=':
    return ['-size', '+' + qual.value + 'c']

def noop_qual(qual):
  return ['-true']

RS = chr(30)
US = '\t'  # chr(31)

BUILTINS = {
  'accessed': ('%A+', time_qual('a'), time_serialize),
  'filename': ('%f', name_qual),
  'changed': ('%C+', time_qual('c'), time_serialize),
  'depth': ('%d', depth_qual),
  'dirname': ('%h', dir_qual),
  'filesystem': ('%F', fs_qual),
  'gid': ('%G', num_qual('gid')),
  'group': ('%g', owner_qual('group')),
  'hardlink_count': ('%n', hardlink_qual),
  'inum': ('%i', num_qual('inum')),
  'modified': ('%T+', time_qual('m'), time_serialize),
  'path': ('%P', path_qual),
  'permissions': ('%m', perm_qual),
  'permissions_expanded': ('%M', perm_qual),
  'size': ('%s', size_qual),
  'symlink': ('%l', symlink_qual),
  'type': ('%Y', type_qual),
  'uid': ('%U', num_qual('uid')),
  'user': ('%u', owner_qual('user'))
  }


FindWrapperTests.py
#!/usr/bin/env python3

from unittest import TestCase, main
from unittest.mock import patch, MagicMock
from io import StringIO

class ForeignDataWrapper:
  def __init__(self, options, columns):
    pass

class multicorn:
  ForeignDataWrapper = ForeignDataWrapper

import sys
sys.modules['multicorn'] = multicorn

from FindWrapper import FindWrapper

ID_BUILTIN = 0
ID_PATTERN = 1
ID_EXECUTABLE = 2

TEST_DIR = '/usr/share/example/'
TEST_ARGS = ['/usr/bin/find', '-O3', '-ignore_readdir_race', TEST_DIR]

class FindWrapperTests(TestCase):

  def test_root_with_slash(self):
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': '/dir/no/slash/' }, columns)
    self.assertEqual(fw._root, '/dir/no/slash/')

  def test_root_without_slash(self):
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': '/dir/no/slash' }, columns)
    self.assertEqual(fw._root, '/dir/no/slash/')

  @patch('FindWrapper.Popen')
  @patch('FindWrapper.PIPE')
  def test_basic(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\t2017-03-22+22:33:15.3646792370\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR }, columns)
    self.assertIn('path', fw._handlers)
    self.assertEqual(fw._handlers['path'][0], ID_BUILTIN)
    self.assertEqual(fw._handlers['path'][1], '%P')
    self.assertIn('modified', fw._handlers)
    self.assertEqual(fw._handlers['modified'][0], ID_BUILTIN)
    self.assertEqual(fw._handlers['modified'][1], '%T+')
    row = next(fw.execute([], columns))
    expected = TEST_ARGS + [ '-printf', '%P\t%T+\n' ]
    # self.assertTrue(mock_popen.called)
    mock_popen.assert_called_with(expected, stdout=mock_pipe)
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:15.3646792370', 'path': 'example.txt' })

  @patch('FindWrapper.Popen')
  @patch('FindWrapper.PIPE')
  def test_multi(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example1.txt\t2017-03-22+22:33:15.3646792370\n' +
                                         'example2.txt\t2017-03-22+22:33:45.3646792370\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'modified': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR }, columns)
    rows = fw.execute([], columns)
    row = next(rows)
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:15.3646792370', 'path': 'example1.txt' })
    row = next(rows)
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:45.3646792370', 'path': 'example2.txt' })

  @patch('FindWrapper.Popen')
  @patch('FindWrapper.PIPE')
  def test_exec(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\t2017-03-22+22:33:15.3646792370\n' +
                                         'text/plain\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'modified': None, 'mime_type': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR,
                       'mime_type': '/usr/bin/file -b -i' }, columns)
    row = next(fw.execute([], columns))
    self.assertEqual(row, { 'modified': '2017-03-22 22:33:15.3646792370',
                            'path': 'example.txt',
                            'mime_type': 'text/plain' })
    expected = TEST_ARGS + [ '-printf', '%P\t%T+\n',
                             '-exec', '/usr/bin/file', '-b', '-i', '{}', ';' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe)

  @patch('FindWrapper.Popen')
  @patch('FindWrapper.PIPE')
  def test_alias(self, mock_pipe, mock_popen):
    process_mock = MagicMock()
    process_attrs = { 'stdout': StringIO('example.txt\t2017-03-22+22:33:15.3646792370\n') }
    process_mock.configure_mock(**process_attrs)
    mock_popen.return_value = process_mock
    columns = { 'path': None, 'ts': None }
    fw = FindWrapper({ 'root_directory': TEST_DIR,
                       'ts': 'modified' }, columns)
    row = next(fw.execute([], columns))
    self.assertEqual(row, { 'ts': '2017-03-22 22:33:15.3646792370',
                            'path': 'example.txt' })
    expected = TEST_ARGS + [ '-printf', '%P\t%T+\n' ]
    mock_popen.assert_called_with(expected, stdout=mock_pipe)

if __name__ == '__main__':
  main()
